pkg/server/api/middleware/authorization.go

```diff
@@ -59,18 +51,18 @@ func (m *authorizationMiddleware) Preprocess(ctx context.Context, methodName str
 		ctx = rpccontext.WithLogger(ctx, rpccontext.Logger(ctx).WithFields(fields))
 	}
 
-	authorizer, ok := m.authorizers[methodName]
-	if !ok {
-		middleware.LogMisconfiguration(ctx, "Authorization misconfigured (method not registered); this is a bug")
-		return nil, status.Errorf(codes.Internal, "authorization misconfigured for %q (method not registered)", methodName)
-	}
-
-	authorizedCtx, err := authorizer.AuthorizeCaller(ctx)
+	ctx, allow, err := m.opaAuth(ctx, req, methodName)
 	if err != nil {
 		rpccontext.Logger(ctx).WithError(err).Error("Failed to authenticate caller")
 		return nil, err
 	}
-	return authorizedCtx, nil
+	if allow {
+		return ctx, nil
+	}
+
+	deniedErr := status.Errorf(codes.PermissionDenied, "Authorization denied for method %s", methodName)
+	rpccontext.Logger(ctx).WithError(deniedErr).Error("Failed to authenticate caller")
```

Comment:
"Failed to authenticate caller" is handled higher up in the function, where we got some error in the opa auth. I think this error log should be "Failed to authorize caller", or maybe better "Caller is unauthorized" since it's not really a failure of the service but instead an explicit rejection.

Reply:
i'm not sure if i understood your comment correctly. i've updated it in https://github.com/spiffe/spire/pull/2416/commits/bdf9ad4736eec98efa955e2d778b3e2b6da7a348, if you can take a quick look.

`Failed to authenticate caller` is what is being passed back today if the authentication middleware denies.

Reply:
The new wording addresses my comment.

The previous logic had a general error for authorization, which could mean either the auth itself failed (could be authn or authz failure) or there was no "error" but instead the caller is rejected but authz. So originally it may have somewhat made sense to bundle all that into a single error return.

With this change, we now explicitly know if there was an error, or if there was no error and the call is authorized or not. An error now is actually a real error, and the true/false boolean is an explicit authz accept or deny return. So for the flag check, there is no "failure" error to log, but instead "you're not authorized". Does that make sense?

Reply:
Ah yes! Gotcha, thanks for providing more detail!